### DISCUSSION TITLE
How to use multi bindings in v2?

### DISCUSSION THREAD

**Discussion by yzqdev** [wailsapp/wails#1499] [Q&A] [answered]
As the document says: In v1, the [runtime](https://wails.top/docs/reference/runtime/intro) was available via a struct passed to WailsInit(). In v2, the runtime has been moved out to its own package. Each method in the runtime takes the context.Context that is passed to the [OnStartup](https://wails.top/docs/reference/options#onstartup) method.

I create another binding `conf.go`, but the context is assigned in the app's `startup` method. only the `app` 's ctx is not null.

in v1,add  many bindings with runtime are easy because there is no ctx needed. Now i wonder how to do  in v2.

### main.go
```go

//go:embed frontend/dist
var assets embed.FS
func main() {
	app := NewApp()
	conf := NewConf()// the new binding
	err := wails.Run(&options.App{
		Title:            "genshin-wails",
		Width:            1024,
		Height:           768,
		Assets:           assets,
		OnStartup:        app.startup,
		Bind: []interface{}{
			app, conf,
		},
	})
	if err != nil {
		println("Error:", err)
	}
}
```
### app.go
```go
type App struct {
	ctx context.Context
}
func NewApp() *App {
	return &App{}
}
// startup is called when the app starts. The context is saved
// so we can call the runtime methods
func (a *App) startup(ctx context.Context) {
	a.ctx = ctx
}
```
### conf.go
```go
type Config struct {
	ctx context.Context
}

func NewConf() *Config {
	return &Config{}
}
//the method below can not run normally
func (c *Config) SetOutDir() string {
	c.ctx = context.Background()// this context is wrong,  it should be the same with app's ctx
	runtime.LogInfo(c.ctx, "log info")
	runtime.BrowserOpenURL(c.ctx, "e:/")
	return "success"
}
```

The method `SetOutDir` will say  `ctx is null` or `An invalid context was passed. This method requires the specific context given in the lifecycle hooks`, so how to assign a context in multi binding? minimal repo https://github.com/yzqbugs/wails-qa

**Comment by leaanthony** [comment, marked as answer]
Good question! There are many ways (as this is just a Go question). For this scenario I would use a function in `OnStartup` to call methods on your structs with the context.

```
OnStartup: func(ctx context.context) {
  app.Init(ctx)
  config.Init(ctx)
},
```

**Reply by yzqdev** [reply]
Awesome! Thank you!😀

**Reply by leaanthony** [reply]
Thanks for asking. I feel this should be in the docs as an example 👍
